File: back-end/ShopHangTet/Controllers/ProductsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using ShopHangTet.DTOs;
using ShopHangTet.Services;

namespace ShopHangTet.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class ProductsController : ControllerBase
    {
        private readonly IProductService _productService;

        public ProductsController(IProductService productService)
        {
            _productService = productService;
        }

        [HttpGet("items")]
        public async Task<IActionResult> GetItems([FromQuery] string? name)
        {
            var items = await _productService.GetItemsAsync(name);
            return Ok(items);
        }

        [HttpGet("gift-boxes")]
        public async Task<IActionResult> GetGiftBoxes([FromQuery] string? name)
        {
            var giftBoxes = await _productService.GetGiftBoxesAsync(name);
            return Ok(giftBoxes);
        }

        [HttpGet("collections")]
        public async Task<IActionResult> GetCollections([FromQuery] string? name)
        {
            var collections = await _productService.GetCollectionsAsync(name);
            return Ok(collections);
        }

        [HttpGet("items/{id}")]
        public async Task<IActionResult> GetItemById(string id)
        {
            var item = await _productService.GetItemByIdAsync(id);
            if (item == null) return NotFound();
            return Ok(item);
        }

        [HttpGet("gift-boxes/{id}")]
        public async Task<IActionResult> GetGiftBoxById(string id)
        {
            var giftBox = await _productService.GetGiftBoxDetailByIdAsync(id);
            if (giftBox == null) return NotFound();
            return Ok(giftBox);
        }

        [HttpGet("collections/{id}")]
        public async Task<IActionResult> GetCollectionById(string id)
        {
            var collection = await _productService.GetCollectionDetailByIdAsync(id);
            if (collection == null) return NotFound();
            return Ok(collection);
        }

        // Admin GiftBox CRUD with auto pricing.

        /// Create GiftBox with calculated price.
        [HttpPost("gift-boxes")]
        [Authorize(Roles = "ADMIN")]
        public async Task<IActionResult> CreateGiftBox([FromBody] CreateGiftBoxDto dto)
        {
            try
            {
                var giftBox = await _productService.CreateGiftBoxAsync(dto);
                return Ok(ApiResponse<object>.SuccessResult(new
                {
                    id = giftBox.Id,
                    name = giftBox.Name,
                    price = giftBox.Price,
                    collectionId = giftBox.CollectionId
                }, "GiftBox created with auto-calculated price"));
            }
            catch (Exception ex)
            {
                return BadRequest(ApiResponse<string>.ErrorResult(ex.Message));
            }
        }

        /// Update GiftBox and recalculate price when needed.
        [HttpPut("gift-boxes/{id}")]
        [Authorize(Roles = "ADMIN")]
        public async Task<IActionResult> UpdateGiftBox(string id, [FromBody] UpdateGiftBoxDto dto)
        {
            try
            {
                var giftBox = await _productService.UpdateGiftBoxAsync(id, dto);
                return Ok(ApiResponse<object>.SuccessResult(new
                {
                    id = giftBox.Id,
                    name = giftBox.Name,
                    price = giftBox.Price,
                    collectionId = giftBox.CollectionId
                }, "GiftBox updated"));
            }
            catch (Exception ex)
            {
                return BadRequest(ApiResponse<string>.ErrorResult(ex.Message));
            }
        }

        /// Preview GiftBox price without saving.
        [HttpPost("gift-boxes/calculate-price")]
        [Authorize(Roles = "ADMIN")]
        public async Task<IActionResult> CalculateGiftBoxPrice([FromBody] CalculateGiftBoxPriceDto dto)
        {
            try
            {
                var items = dto.Items.Select(i => new ShopHangTet.Models.GiftBoxItem
                {
                    ItemId = i.ItemId,
                    Quantity = i.Quantity
                }).ToList();

                var price = await _productService.CalculateGiftBoxPriceAsync(dto.CollectionId, items);
                return Ok(ApiResponse<object>.SuccessResult(new
                {
                    collectionId = dto.CollectionId,
                    calculatedPrice = price
                }, "Price calculated from collection pricing rule"));
            }
            catch (Exception ex)
            {
                return BadRequest(ApiResponse<string>.ErrorResult(ex.Message));
            }
        }
    }
}

File: back-end-nodejs/src/controllers/productsController.js
import { Router } from 'express';
import { ApiResponse } from '../dtos/index.js';
import { authenticate, authorize } from '../middlewares/index.js';

/**
 * Products Controller - Tương đương ProductsController.cs
 * Routes: /api/products
 */
export function createProductsRouter(productService) {
  const router = Router();

  // ========== GET /api/products/items ==========
  router.get('/items', async (req, res) => {
    try {
      const items = await productService.getItems(req.query.name);
      return res.status(200).json(items);
    } catch (error) {
      console.error('GetItems error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ========== GET /api/products/gift-boxes ==========
  router.get('/gift-boxes', async (req, res) => {
    try {
      const giftBoxes = await productService.getGiftBoxes(req.query.name);
      return res.status(200).json(giftBoxes);
    } catch (error) {
      console.error('GetGiftBoxes error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ========== GET /api/products/collections ==========
  router.get('/collections', async (req, res) => {
    try {
      const collections = await productService.getCollections(req.query.name);
      return res.status(200).json(collections);
    } catch (error) {
      console.error('GetCollections error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ========== GET /api/products/items/:id ==========
  router.get('/items/:id', async (req, res) => {
    try {
      const item = await productService.getItemById(req.params.id);
      if (!item) return res.status(404).json(ApiResponse.error('Item not found'));
      return res.status(200).json(item);
    } catch (error) {
      console.error('GetItemById error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ========== GET /api/products/gift-boxes/:id ==========
  router.get('/gift-boxes/:id', async (req, res) => {
    try {
      const giftBox = await productService.getGiftBoxDetailById(req.params.id);
      if (!giftBox) return res.status(404).json(ApiResponse.error('GiftBox not found'));
      return res.status(200).json(giftBox);
    } catch (error) {
      console.error('GetGiftBoxById error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ========== GET /api/products/collections/:id ==========
  router.get('/collections/:id', async (req, res) => {
    try {
      const collection = await productService.getCollectionDetailById(req.params.id);
      if (!collection) return res.status(404).json(ApiResponse.error('Collection not found'));
      return res.status(200).json(collection);
    } catch (error) {
      console.error('GetCollectionById error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ========== POST /api/products/gift-boxes (ADMIN) ==========
  router.post('/gift-boxes', authenticate, authorize('ADMIN'), async (req, res) => {
    try {
      const giftBox = await productService.createGiftBox(req.body);
      return res.status(200).json(
        ApiResponse.success(
          {
            id: giftBox._id.toString(),
            name: giftBox.name,
            price: giftBox.price,
            collectionId: giftBox.collectionId,
          },
          'GiftBox created with auto-calculated price'
        )
      );
    } catch (error) {
      return res.status(400).json(ApiResponse.error(error.message));
    }
  });

  // ========== PUT /api/products/gift-boxes/:id (ADMIN) ==========
  router.put('/gift-boxes/:id', authenticate, authorize('ADMIN'), async (req, res) => {
    try {
      const giftBox = await productService.updateGiftBox(req.params.id, req.body);
      return res.status(200).json(
        ApiResponse.success(
          {
            id: giftBox._id.toString(),
            name: giftBox.name,
            price: giftBox.price,
            collectionId: giftBox.collectionId,
          },
          'GiftBox updated'
        )
      );
    } catch (error) {
      return res.status(400).json(ApiResponse.error(error.message));
    }
  });

  // ========== POST /api/products/gift-boxes/calculate-price (ADMIN) ==========
  router.post('/gift-boxes/calculate-price', authenticate, authorize('ADMIN'), async (req, res) => {
    try {
      const { CollectionId, Items } = req.body;

      const items = (Items || []).map((i) => ({
        itemId: i.ItemId,
        quantity: i.Quantity,
      }));

      const price = await productService.calculateGiftBoxPrice(CollectionId, items);
      return res.status(200).json(
        ApiResponse.success(
          {
            collectionId: CollectionId,
            calculatedPrice: price,
          },
          'Price calculated from collection pricing rule'
        )
      );
    } catch (error) {
      return res.status(400).json(ApiResponse.error(error.message));
    }
  });

  return router;
}
